concrete-boolean/src/engine/bootstrapping/cpu.rs

```diff
@@ -130,23 +64,104 @@ impl Memory {
     }
 }
 
+#[cfg(feature = "fftw")]
+type FourierLweBootstrapKey32 = FftwFourierLweBootstrapKey32;
+#[cfg(feature = "fftw")]
+type FourierEngine = FftwEngine;
+
+#[cfg(all(feature = "fft", not(feature = "fftw")))]
```

Comment:
if a user activates both the fft and fftw features, they only get the fftw backend. i'm not sure why we want this to be the case.

Reply:
The idea of this, is that `fft` is the default feature (so does not needs to be enabled by the user), so if the user enables the `fftw` feature  it means the user really wants fftw, so we only use the fftw)

A bit like the cuda backend, where by default we compile to do backend on CPU, but if the user compiles with de `cuda` 
feature we compile to do the bootstrap on GPU.

The other option would have been to not make a default choice and let the user explicitely chose the fft backend / cuda.

I think the fftw backend is likely to simply go away as its less performant, less portable, longer to compile,
so I choose the first option.

Do you have other preferences on the way to expose this feature choice ?

Reply:
the problem is that the user isn't the only one that decides which features they want.

let's say the user is writing application A, and depends on concrete and some other library B  
library B also depends on concrete. if library B enables the fftw feature, then the feature gets enabled automatically for all the other libraries in the build tree (in this case, application A).

so this means that a dependency can enable fftw for you without your knowledge.

i'm not sure what the best solution is

Reply:
The same thing happens with the cuda feature.

A 'clean' solution would be to make these features purely additive, and add a runtime option to the ServerKey
that lets user select the backend to use.


```Rust

pub enum Backend {
#[cfg(feature = "fft")]
Fft,
#[cfg(feature = "fftw")]
Fftw,
#[cfg(feature = "cuda")]
Cuda
}
```

But then we need to select a sensible default cuda > fft > fftw, and make sure at least one backend is enabled at compile time

Reply:
i think i personally prefer this solution. then we can use a runtime flag or environment variable to set the backend, which gives the end user more freedom. but maybe this is better left for a separate PR. what do you think?

Reply:
I prefer a flag set at runtime on the ServerKey rather than an environment variable.

So, you propose that we merge this as is, then in a later PR add the "runtime backend selection" ?

Reply:
yeah, i think it'll be simpler to do it that way. since we already have the same issue with the cuda backend and solving it is outside the goals of this PR